### ISSUE TITLE
Interface design for timeout

### ISSUE THREAD

**Issue by matrixji** [milvus-io/milvus-sdk-cpp#49]
Some interfaces have timeout options. e.g.
```cpp
    /**
     * Load specific partitions data of one collection into query nodes.
     * If the timeout is specified, this api will call ShowPartitions() to check partition's loading state,
     * waiting until the collection completely loaded into query node.
     *
     * @param [in] collection_name name of the collection
     * @param [in] partition_names name array of the partitions
     * @param [in] timeout timeout setting for loading, set to nullptr to return instantly
     * @return Status operation successfully or not
     */
    virtual Status
    LoadPartitions(const std::string& collection_name, const std::vector<std::string>& partition_names,
                   const TimeoutSetting* timeout) = 0;
```

So, some points from my side:
- As the return is Status, do we need to add a TIMEOUT StatusCode for the timeout scenario?
- As timeout is read-only, I'd like to use `const TimeoutSettings &` instead of const Timeout*, but without c++17 we don't have std::optional, so I suggest split this interface to 2 functions as below:
  ```cpp
   Status LoadPartitions(const std::string&, const std::vector<std::string>&);
   Status LoadPartitions(const std::string&, const std::vector<std::string>&, const TimeoutSetting&);
  ```
- Discussion, as we use ShowPartitions to check the loading state, and internally we have the percentage of loaded, do we need to introduce some callback in this interface, then users could use it as percentage reporting in their application? Like:
  ```cpp
  Status LoadPartitions(const std::string&, const std::vector<std::string>&, const TimeoutSetting&,
                        std::function<void(const CollectionInfo&)>);
  ```
- Discussion continue, developers want to write an application without BLOCK, Right now the API is Sync, but most of them is not blocking(except the RPC call), and the timeout options will let the API call in a BLOCK state in code logic until it succeed or timeout, so do we need to provide some interface using std::async? Like:
  ```cpp
  std::future<Status> LoadPartitions(const std::string&, const std::vector<std::string>&, const TimeoutSetting&);
  std::future<Status> LoadPartitions(const std::string&, const std::vector<std::string>&, const TimeoutSetting&,
                                     std::function<void(const CollectionInfo&)>);
  ```

**Comment by yhmo** [comment]
1. Right, a status code for timeout is necessary.
2. I also prefer "`const TimeoutSetting&`". The point I use "`const TimeoutSetting*`" is: if the client want to ignore timeout, he can pass `nullptr`, no need to construct a TimeoutSetting object.  But I accept this is not a strong reason to use `const TimeoutSetting*`, we can ask client always pass a TimeoutSetting object.
3. I think the callback function can be as a member of TiemoutSetting class.
4. Most of interfaces are blocking call(by rpc), except the flush()/LoadCollection()/LoadPartitions()/GetCollectionStatistics()/GetPartitionStatistics(), maybe the ManualCompaction() also can be async. Two options here: 1) only these async interfaces return `std::future<Status>`, others return `Status`. 2) provide async interfaces with different function name, for example: `std::future<Status> LoadPartitionsAsync(...)`. Which one is better?

**Comment by matrixji** [comment]
According to comments from yhmo:

Interface updates to:
```
Status LoadPartitions(const std::string&, const std::vector<std::string>&, const TimeoutSetting&);
std::future<Status> LoadPartitionsAsync(const std::string&, const std::vector<std::string>&, const TimeoutSetting&);
```
- Add TIMEOUT in StatusCode
- Add callback in TimeoutSetting object, by default no oper.
- Add TimeoutSetting::NoWait() for return a object for without waiting

@ArkaprabhaChakraborty do you have any comments about it, I see you are handling the LoadCollections interface, they are similar.

**Comment by ArkaprabhaChakraborty** [comment]
Yes even I thought about it. The LoadCollections() is definitely similar and I think we need a timeout parameter. The status code can have an enum like it has for all and InMemory.

**Comment by ArkaprabhaChakraborty** [comment]
I prefer @yhmo second point about the timeout parameter setting.

**Comment by yhmo** [comment]
> According to comments from yhmo:
> 
> Interface updates to:
> 
> ```
> Status LoadPartitions(const std::string&, const std::vector<std::string>&, const TimeoutSetting&);
> std::future<Status> LoadPartitionsAsync(const std::string&, const std::vector<std::string>&, const TimeoutSetting&);
> ```
> 
> * Add TIMEOUT in StatusCode
> * Add callback in TimeoutSetting object, by default no oper.
> * Add TimeoutSetting::NoWait() for return a object for without waiting
> 
> @ArkaprabhaChakraborty do you have any comments about it, I see you are handling the LoadCollections interface, they are similar.

If the client input a  TimeoutSetting::NoWait() to the LoadPartitionsAsync(), it ought to return Status::OK() right?

**Comment by yhmo** [comment]
I get two points, need us discuss and make decision. 

1) In the python sdk, each interface has a "timeout" parameter: https://github.com/milvus-io/pymilvus/blob/master/pymilvus/client/stub.py.
   Maybe we need add a timeout parameter for each C++ sdk interface? Or declare async method for each interface?
   We can use `grpc::ChannelArguments` to set `GRPC_ARG_KEEPALIVE_TIME_MS` for a rpc channel, but this value is not proper for timeout.

2) We add callback to report percentage of a progress, for the LoadCollection()/LoadPartitions()/CreateIndex(). 
   Since we want to put the callback into TimeoutSetting. The call backfucntion shoule be general. I think we can define the callback like this:
`std::function<void(int, int)> func_;`
The first int value is how many items is finished, the second int value is totally how many items there. We also can define a struct for this:
```
        struct Progress {
          int finished_;
          int total_;
        }
       std::function<void(Progress&)> func_;
```
 Do you think it make sense?

**Comment by ArkaprabhaChakraborty** [comment]
I think this will work.

**Comment by yhmo** [comment]
> I think this will work.

For the point 1, I am not sure whether it is worth to do(add a timeout parameter for each C++ sdk interface Or declare async method for each interface). Personally I intend to keep the code simple. Maybe can take it as enhancement to do in future.

**Comment by matrixji** [comment]
- For the grpc timeout, I guess it's possible to introduce a timeout option for MilvusClient (not for each interface).
- For the async API, maybe it's too complicated right now, sync API is easier. For the advanced users, I guess they prefer to use grpc directly, the grpc has CompletionQueue for handling async things. So currently I'm not implementing the `LoadPartitionsAsync`, seems we need more discussion or more comments from the community.
- For the callback object, currently,IMO It's OK for adding it in TimeoutSetting, maybe  we need to rename the class, it includes:
  - query status interval
  - query status timeout
  - progress callback

**Comment by yhmo** [comment]
> * For the grpc timeout, I guess it's possible to introduce a timeout option for MilvusClient (not for each interface).
> * For the async API, maybe it's too complicated right now, sync API is easier. For the advanced users, I guess they prefer to use grpc directly, the grpc has CompletionQueue for handling async things. So currently I'm not implementing the `LoadPartitionsAsync`, seems we need more discussion or more comments from the community.
> * For the callback object, currently,IMO It's OK for adding it in TimeoutSetting, maybe  we need to rename the class, it includes:
>   
>   * query status interval
>   * query status timeout
>   * progress callback

Yes, the grpc has generated async functions in the Stub, useful for future async functions.

Actually the LoadCollection()/LoadPartitions() we discussed is sync function, client call the function and wait until it finish. So we don't need to create another function named LoadCollectionAsync(), reserve this name for future real-async functions.

I can send a message in milvus-users channel in slack, to notify we are working on this project, I will ask user who is interest in this to join the milvus-sdk-cpp channel, give us usage suggestion and comment.

Then how about rename the TimeoutSetting to be ProgressMonitor?

**Comment by matrixji** [comment]
Yes, I agree ProgressMonitor is more accurate.

**Comment by ArkaprabhaChakraborty** [comment]
> Then how about rename the TimeoutSetting to be ProgressMonitor?
I guess that would be better.

**Comment by yhmo** [comment]
Okey, let me make a commit to update code according to our conclusion.
List conclusion here:
- In this stage we only implement sync interfaces.
- Introduce a timeout option for MilvusClient (not for each interface), we can defer this to next stage when we start implement async interfaces.
- Rename TimeoutSetting to be ProgressMonitor
- Add callback in ProgressMonitor object, by default no oper
- The definition of ProgressMonitor:
    ```
   struct Progress {
          int finished_;
          int total_;
    }
       
   class ProgressMonitor {
   private:
       int64_t check_interval_ = 0;
       int64_t check_timeout_ = 0;  // set to 0 for NoWait()
       std::function<void(Progress&)> callback_func_;
   }
```
- Add ProgressMonitor::NoWait() for return a object for without waiting
- Add TIMEOUT in StatusCode, return this error when the ProgressMonitor is time out, also returned by async interfaces(next stage)
- So, the related interfaces should be like:
  `Status LoadPartitions(const std::string&, const std::vector<std::string>&, const ProgressMonitor&);`

Anything I missed?
